### PR TITLE
Some modifications for better allocation

### PR DESCRIPTION
**Pull request by Sonata165**
- Modified some variable name and added some comments for readability. 
- Some students are lost in the final results. Solved after adding completeness term in fitness function.
- There are usually groups with undesired size. Solved after adding group size penalty term to fitness function.
- (Not finished) There are too many common topics. Can consider the similarity of topic interest inside the fitness function.

### CODE REVIEW

**Comment by mrqorib** [comment]
Hi Longsheng, thank you for your pull request. Could you please:
1. Not committing your actual configuration file, so that the config file is more understandable to other people who want to use the code for other purposes
2. Explain which part of the original code loses some students, and how you fixed it. I think this doesn't happen with me and Taha's experience
3. Explain what you meant by undesired group sizes, is it < min_member or > max_member?
4. Explain why you need a separate script to validate the group allocation (validate_alloc_result.py), I think it's better to validate it during the optimization so that users do not need to run the script multiple times.
In addition, I think for now it's better to not move the code to the src folder so that Git can automatically check the differences.

Thanks!

**Comment by Sonata165** [comment]
1. Deleted 4248.json (also other unnecessary files)
2. I don't know which part results in losing students. When I run the allocation, we have 180 students in total, but there is usually less than 170 students in the allocation results file before this modification.
3. I set desired group size as 5-6 by min_member and max_member in argparse, but there are usually smaller groups in the allocation results (2-3 size group) before this modification.
4. The results might be validated with different rules when it's used by different people (I just deleted the file this time)

Thank you.

File: form_group.py [hunk 1]
```diff
@@ -90,171 +117,209 @@ def merge_group(subgroups, score_only=True):
...
+        # We need at least one person has certain level of programming skill
+        max_prog = max([int(i) for i in comb_skills['prog']])
+        if max_prog < config['skills']['prog']['min']:  # Longshen: original is <=
             score = 0
         elif min_topic_diff is None:
             score = 0
 
+        # If group size undesired, add penalty.
+        group_size = len(merged_group['id'])
+        if 5 <= group_size <= 6:
```
Comment (mrqorib) [inline form_group.py]: Could you please change the numbers to min_member and max_member?

**Review by mrqorib — CHANGES_REQUESTED** [review]
Thanks a lot, @Sonata165! I'm glad you understand the code well amid the lack of documentation and bad variable naming. Thanks for fixing the variable naming! I agree that it's indeed a strange way to initialize the chromosomes haha, I welcome other methods of initialization.

I only have a small comment about the numbers. I also think it's better to remove reza_notes.txt and put the explanation in the README. Let me know if you want me to do it. 👍

File: form_group.py (hunk 1 above)
Comment (Sonata165) [reply form_group.py]: Oh sure, sorry for the late reply. Just modified accordingly.

**Review by mrqorib — APPROVED** [review]
Thank you for your help!

**Pull request merged**